filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,526 @@
+		return err
+	}
+}
+
+func WithEnvoyMetadata(metadata *ext_authz_v3_core.Metadata) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = metadata
+		return nil
+	}
+}
+
+func WithEnvoyMetadataBytes(content []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyMetadata = &ext_authz_v3_core.Metadata{}
+
+		return protojson.Unmarshal(content, cfg.envoyMetadata)
+	}
+}
+
+func WithEnvoyMetadataFile(file string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		content, err := os.ReadFile(file)
+		if err != nil {
+			return err
+		}
+
+		err = WithEnvoyMetadataBytes(content)(cfg)
+		if err != nil {
+			return fmt.Errorf("cannot parse '%v': %w", file, err)
+		}
+
+		return nil
+	}
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyMetadata() *ext_authz_v3_core.Metadata {
+	return cfg.envoyMetadata
+}
+
+func NewOpenPolicyAgentConfig(opts ...func(*OpenPolicyAgentInstanceConfig) error) (*OpenPolicyAgentInstanceConfig, error) {
+	cfg := OpenPolicyAgentInstanceConfig{}
+
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	if cfg.configTemplate == nil {
+		var err error
+		cfg.configTemplate, err = os.ReadFile("opaconfig.yaml")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cfg, nil
+}
+
+func (registry *OpenPolicyAgentRegistry) Close() {
+	registry.once.Do(func() {
+		registry.mu.Lock()
+		defer registry.mu.Unlock()
+
+		ctx, cancel := context.WithTimeout(context.Background(), defaultShutdownGracePeriod)
+		defer cancel()
+
+		for _, instance := range registry.instances {
+			instance.Close(ctx)
+		}
+
+		registry.closed = true
+		close(registry.quit)
+	})
+}
+
+func (registry *OpenPolicyAgentRegistry) cleanUnusedInstances(t time.Time) {
+	registry.mu.Lock()
+	defer registry.mu.Unlock()
+
+	if registry.closed {
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultShutdownGracePeriod)
+	defer cancel()
+
+	for key, inst := range registry.instances {
+		lastused, ok := registry.lastused[inst]
+
+		if ok && t.Sub(lastused) > registry.reuseDuration {
+			inst.Close(ctx)
+
+			delete(registry.instances, key)
+			delete(registry.lastused, inst)
+		}
+	}
+}
+
+func (registry *OpenPolicyAgentRegistry) startCleanerDaemon() {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-registry.quit:
+			return
+		case t := <-ticker.C:
+			registry.cleanUnusedInstances(t)
+		}
+	}
+}
+
+// Do implements routing.PostProcessor and cleans unused OPA instances
+func (registry *OpenPolicyAgentRegistry) Do(routes []*routing.Route) []*routing.Route {
+	rr := make([]*routing.Route, len(routes))
+	inUse := make(map[*OpenPolicyAgentInstance]struct{})
+
+	for i, ri := range routes {
+		rr[i] = ri
```

Comment:
Do we need a copy here?

Reply:
I don't think so. We can also just return `routes`. This is "inspired" by https://github.com/zalando/skipper/blob/fa84c771951b0ed891c31f36d4ffd45c094eed0d/scheduler/scheduler.go#L582 where is is done similarly.